modules/beefy/src/lib.rs

```diff
@@ -0,0 +1,651 @@
+	}
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
+
+	#[pallet::hooks]
+	impl<T: Config<I>, I: 'static> Hooks<BlockNumberFor<T>> for Pallet<T, I> {
+		fn on_initialize(_n: T::BlockNumber) -> frame_support::weights::Weight {
+			<RequestCount<T, I>>::mutate(|count| *count = count.saturating_sub(1));
+
+			Weight::from_ref_time(0)
+				.saturating_add(T::DbWeight::get().reads(1))
+				.saturating_add(T::DbWeight::get().writes(1))
+		}
+	}
+
+	impl<T: Config<I>, I: 'static> OwnedBridgeModule<T> for Pallet<T, I> {
+		const LOG_TARGET: &'static str = LOG_TARGET;
+		type OwnerStorage = PalletOwner<T, I>;
+		type OperatingMode = BasicOperatingMode;
+		type OperatingModeStorage = PalletOperatingMode<T, I>;
+	}
+
+	#[pallet::call]
+	impl<T: Config<I>, I: 'static> Pallet<T, I>
+	where
+		BridgedMmrHashing<T, I>: 'static + Send + Sync,
+	{
+		/// Initialize pallet with BEEFY authority set and best known finalized block number.
+		#[pallet::weight((T::DbWeight::get().reads_writes(2, 3), DispatchClass::Operational))]
+		pub fn initialize(
+			origin: OriginFor<T>,
+			init_data: InitializationDataOf<T, I>,
+		) -> DispatchResult {
+			Self::ensure_owner_or_root(origin)?;
+
+			let is_initialized = <ImportedCommitmentsInfo<T, I>>::exists();
+			ensure!(!is_initialized, <Error<T, I>>::AlreadyInitialized);
+
+			log::info!(target: LOG_TARGET, "Initializing bridge BEEFY pallet: {:?}", init_data);
+			Ok(initialize::<T, I>(init_data)?)
+		}
+
+		/// Change `PalletOwner`.
+		///
+		/// May only be called either by root, or by `PalletOwner`.
+		#[pallet::weight((T::DbWeight::get().reads_writes(1, 1), DispatchClass::Operational))]
+		pub fn set_owner(origin: OriginFor<T>, new_owner: Option<T::AccountId>) -> DispatchResult {
+			<Self as OwnedBridgeModule<_>>::set_owner(origin, new_owner)
+		}
+
+		/// Halt or resume all pallet operations.
+		///
+		/// May only be called either by root, or by `PalletOwner`.
+		#[pallet::weight((T::DbWeight::get().reads_writes(1, 1), DispatchClass::Operational))]
+		pub fn set_operating_mode(
+			origin: OriginFor<T>,
+			operating_mode: BasicOperatingMode,
+		) -> DispatchResult {
+			<Self as OwnedBridgeModule<_>>::set_operating_mode(origin, operating_mode)
+		}
+
+		/// Submit a commitment generated by BEEFY authority set.
+		///
+		/// It will use the underlying storage pallet to fetch information about the current
+		/// authority set and best finalized block number in order to verify that the commitment
+		/// is valid.
+		///
+		/// If successful in verification, it will update the underlying storage with the data
+		/// provided in the newly submitted commitment.
+		#[pallet::weight(0)]
+		pub fn submit_commitment(
+			origin: OriginFor<T>,
+			commitment: BridgedBeefySignedCommitment<T, I>,
+			validator_set: BridgedBeefyAuthoritySet<T, I>,
+			mmr_leaf: Box<BridgedBeefyMmrLeaf<T, I>>,
+			mmr_proof: BridgedMmrProof<T, I>,
+		) -> DispatchResult
```

Comment:
I think in the first version we deploy we should just save the current `validator_set` (full set of BEEFY validator keys) in storage instead of using block transaction space here on each call.

We know from MMR Leaf contents on which block the set changes and have its "metadata" (validator set keys mmr root) to validate the input.

This call will then only accept a `commitment` if either:
- it's for a BEEFY mandatory block (session-boundary block) _and_ full new validator set is also passed as argument (after validating it against known metadata, we update storage with new set),
- or, it's for a non-mandatory BEEFY block, _and_ `validator_set` param is `None` and the one from storage is used (we also verify the one from storage matches current commitment metadata).

```suggestion
		pub fn submit_commitment(
			origin: OriginFor<T>,
			commitment: BridgedBeefySignedCommitment<T, I>,
			validator_set: Option<BridgedBeefyAuthoritySet<T, I>>,
			mmr_leaf: Box<BridgedBeefyMmrLeaf<T, I>>,
			mmr_proof: BridgedMmrProof<T, I>,
		) -> DispatchResult
```

or we could expose different `submit_mandatory_commitment()` call that takes `validator_set` while `submit_commitment()` doesn't.

Reply:
We can do this. The current approach is not efficient at all so we'll definitely need to change it before deploying. Would it be ok to track this as part of #1608 ?